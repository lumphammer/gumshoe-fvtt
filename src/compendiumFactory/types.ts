import { generalAbility, investigativeAbility } from "../constants";

type BaseAbilityTemplate = {
  category: string,
  hasSpecialities: boolean,
  specialities: string[],
  rating: number,
  pool: number,
  min: number,
  max: number,
  occupational: boolean,
  showTracker: boolean,
  img: string,
}

export type InvestigativeAbilityTemplate = {
  type: typeof investigativeAbility,
} & BaseAbilityTemplate;

export type GeneralAbilityTemplate = {
  type: typeof generalAbility,
  canBeInvestigative: boolean,
  refreshesDaily: boolean,
} & BaseAbilityTemplate;

type Name = {
  name: string,
};

export type BaseAbilityDump<T> = {
  [category: string]: Array<Name & Partial<T>>,
};

export type InvestigativeAbilitiesData = BaseAbilityDump<InvestigativeAbilityTemplate>;

export type GeneralAbilitiesData = BaseAbilityDump<GeneralAbilityTemplate>;
