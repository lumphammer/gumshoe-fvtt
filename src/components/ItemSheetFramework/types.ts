export enum ItemSheetMode {
  Main = "main",
  Config = "config",
}
