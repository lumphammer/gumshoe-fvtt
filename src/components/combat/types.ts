export type TurnInfo = {
  id: string;
  name: string;
  img: string;
  active: boolean;
  defeated: boolean;
  hidden: boolean;
  initiative: number | undefined | null;
  resource: number | boolean | string | null;
  effects: foundry.documents.ActiveEffect[];
  passingTurnsRemaining: number;
  totalPassingTurns: number;
};
