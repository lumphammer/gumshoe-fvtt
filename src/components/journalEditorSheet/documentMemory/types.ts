import { Change } from "textdiff-create";

/**
 * An edit to a document. The change itself is a diff from textdiff-create, and
 * we also store the timestamp of the edit.
 */
export interface Edit {
  /** The actual change, as a diff */
  changes: Change[];
  /** The timestamp of the edit, in epoch seconds */
  timestamp: number;
  /** The serial number of the edit, unique for the document */
  serial: number;
}

/**
 * A stack of edits in a document memory. Each stack represents an older stage
 * of the memory and has pointer the next, even, older, one. So, the freshest
 * edits are in the first stack. Every time that fills once, a  combnined edit
 * isadded to the next stack, recursively.
 */
export interface Stack {
  /** The edits in this stack. Will always have length == period of the parent
   * memory
   */
  edits: Array<Edit>;
  /**
   * The state of the document the last time this stack pushed down to the
   * next stack.
   */
  bombBay: Array<Edit>;
  /**
   * A pointer to the next stack, to which we will dump edits every time we have
   * recorded `period` edits.
   */
  next: Stack | null;
}

/**
 * A document's edit memory. Older edits are stored with less granularity.
 * Any commit can be restored by walking the stacks from the start.
 */
export interface BareDocumentMemory {
  /**
   * A pointer to the first stack in the memory, the one which will contain the
   * most recent edits.
   */
  stack: Stack;
  /**
   * The curent serial number.
   */
  serial: number;
  /**
   * The period of this memory, i.e. how many edits are stored in each stack.
   */
  period: number;
  /**
   * Limit on maximum depth of stacks. The total number of edits stored will be
   * period * maxDepth, representing
   * period^maxDepth + period^(maxDepth - 1) + ... + period^1
   * total edits.
   */
  maxDepth: number | null;
}

/**
 * A document's edit memory. Older edits are stored with less granularity.
 * Contains snapshots of each stack to accelerate pushes and restores.
 */
export interface DocumentMemory extends BareDocumentMemory {
  /**
   * a cache of snapshots of each level of the document
   */
  snapshots: string[];
  /**
   * A snapshot of the most recent state of the document
   */
  state: string;
}

/**
 * This type is used in to store the document memories globally. The key is the
 * id of the document (aka JournalEntryPage) plus the id of the and the
 */
export type DocumentMemoryCollection = Record<string, BareDocumentMemory>;
