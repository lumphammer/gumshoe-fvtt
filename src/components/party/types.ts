import { AbilityDataSource, AbilityType } from "../../types";

export const typeHeaderKey = "typeHeader" as const;
export const categoryHeaderKey = "categoryHeader" as const;
export const abilityRowkey = "abilityRowString" as const;
export type TypeHeaderData = {
  rowType: typeof typeHeaderKey;
  abilityType: AbilityType;
};
export type CategoryHeaderData = {
  rowType: typeof categoryHeaderKey;
  category: string;
};
export type ActorAbilityInfo = {
  actorId: string;
  abilityId: string | undefined;
  rating: number | undefined;
};
export type AbilityRowData = {
  rowType: typeof abilityRowkey;
  abilityDataSource: AbilityDataSource;
  actorInfo: {
    [actorId: string]: ActorAbilityInfo;
  };
  total: number;
};
export type RowData = TypeHeaderData | CategoryHeaderData | AbilityRowData;
export const isTypeHeader = (data: RowData): data is TypeHeaderData =>
  data.rowType === typeHeaderKey;
export const isCategoryHeader = (data: RowData): data is CategoryHeaderData =>
  data.rowType === categoryHeaderKey;
