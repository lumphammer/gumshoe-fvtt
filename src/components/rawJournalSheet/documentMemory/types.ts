import { Change } from "textdiff-create";

/**
 * An edit to a document. The change itself is a diff from textdiff-create, and
 * we also store the timestamp of the edit.
 */
export interface Edit {
  /** The actual change, as a diff */
  changes: Change[];
  /** The timestamp of the edit, in epoch seconds */
  timestamp: number;
}

/**
 * A stack of edits in a document memory. Each stack represents an older stage
 * of the memory and has pointer the next, even, older, one. So, the freshest
 * edits are in the first stack. Every time that fills once, a  combnined edit
 * isadded to the next stack, recursively.
 */
export interface Stack {
  /** The edits in this stack. Will always have length == period of the parent
   * memory
   */
  edits: Array<Edit | null>;
  /**
   * The state of the document the last time this stack was pushed to. We store
   * this so that we can calculate a diff when a new state comes in. The
   * alternative being to store reversible diffs, i.e. diffs that contain the
   * actual text from deletions so they can be reversed, which could make diffs
   * very large if there are lots of large changes.
   * XXX can we do anything cleverer?
   */
  // lastPushSnapshot: string;
  /**
   * The state of the document the last time this stack pushed down to the
   * next stack.
   */
  bombBay: Change[];
  /**
   * A pointer to the next stack, to which we will dump edits every time we have
   * recorded `period` edits.
   */
  next: Stack | null;
}

/**
 * A document's edit memory. Older edits are stored with less granularity.
 */
export interface DocumentMemory {
  /**
   * A pointer to the first stack in the memory, the one which will contain the
   * most recent edits.
   */
  stack: Stack;
  /**
   * The curent serial number.
   */
  serial: number;
  /**
   * The period of this memory, i.e. how many edits are stored in each stack.
   */
  period: number;
  /**
   * Limit on maximum depth of stacks. The total number of edits stored will be
   * period * maxDepth, representing
   * period^maxDepth + period^(maxDepth - 1) + ... + period^1
   * total edits.
   */
  maxDepth: number | null;
  /**
   * A snapshot of the most recent state of the document
   */
  state: string;
  /**
   * a cache of snapshots of each level of the document
   */
  snapshots: string[];
}
