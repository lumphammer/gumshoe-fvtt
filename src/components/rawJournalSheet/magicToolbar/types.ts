import { ReactNode } from "react";

export type MagicToolbarContentEntry = {
  category: string;
  content: ReactNode;
};

export type MagicToolbarContent = Record<string, MagicToolbarContentEntry>;

export type MagicToolBarRegister = (
  id: string,
  content: MagicToolbarContentEntry,
) => void;

export type MagicToolBarUnregister = (id: string) => void;

export type MagicToolbarRegisterContextDataValue = {
  register: MagicToolBarRegister;
  unregister: MagicToolBarUnregister;
};
