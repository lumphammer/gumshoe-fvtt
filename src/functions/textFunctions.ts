import { escape as escapeText } from "html-escaper";

import { NoteFormat } from "../types";
import { memoizeNullaryOnce } from "./utilities";

// /////////////////////////////////////////////////////////////////////////////
// Toolbox
// /////////////////////////////////////////////////////////////////////////////

// allow bundle splitting of turndown
const makeTurndown = memoizeNullaryOnce(async () => {
  const { default: TurndownService } = await import("turndown");
  class SafeTurndownService extends TurndownService {
    // by default, turndown escapes anything that looks like a markdown control
    // character (like `[` or `#`). I see the logic in this, but actually I feel
    // like for our purposes, if you have markdown-like html, you actually want to
    // have that seen as markdown after you convert? Also, it knackers foundry
    // link codes (which can be rectified, admittedly) and when we use turndown to
    // generate "plain text" we certainly don't want anything escaped.
    escape(text: string) {
      return text;
    }
  }
  // actual turndownservice object for later use
  const turndownService = new SafeTurndownService();
  return turndownService;
});

// allow bundle splitting of xss
const createXss = memoizeNullaryOnce(async () => {
  const {
    FilterXSS,
    whiteList: defaultXssWhitelist,
    escapeAttrValue,
  } = await import("xss");

  // build a custom shitelist for xss that adds "style" to the allowed attributes
  // for everything
  const newWhitelist = Object.fromEntries(
    Object.entries(defaultXssWhitelist).map(([tag, attrList = []]) => [
      tag,
      [...attrList, "style"],
    ]),
  );

  // copilot said this but it does not work
  // newWhitelist["*"] = ["style"];

  // custom xss to allow style attributes and allow images with src attributes.
  // Yes, it's not ideal XSS, but then again this is a collaborative, trusted
  // environment.
  const xss = new FilterXSS({
    whiteList: newWhitelist,
    onTagAttr: function (tag, name, value, isWhiteAttr) {
      if (tag === "img" && name === "src") {
        // escape its value using built-in escapeAttrValue function
        return name + '="' + escapeAttrValue(value) + '"';
      }
    },
  });
  return xss;
});

async function enrichHtml(originalHtml: string): Promise<string> {
  if (typeof TextEditor !== "undefined") {
    const newHtml = await TextEditor.enrichHTML(originalHtml, {
      // @ts-expect-error foundry types don't know about `async` yet
      async: true,
    });
    return newHtml;
  } else {
    return originalHtml;
  }
}

// /////////////////////////////////////////////////////////////////////////////
// Converters
// There are no plaintext <=> markdown converters because it's a no-op.
// /////////////////////////////////////////////////////////////////////////////

async function htmlToMarkdown(html: string) {
  // ever first-time a cool regex and then realise you don't need it, but
  // you're so proud of your ninja regex skills dating back to doing Perl in the
  // 90s that you want to leave it in as a comment? Anyway check this bad boy
  // out. Turns out I don't need it because I've overridden turndown to stop it
  // doing any escaping at all.
  // return turndown(source).replace(/@(\w+)\\\[(\w+)\\\]\{([^}]*)\}/g, (m, p1, p2, p3) => `@${p1}[${p2}]{${p3}}`);
  const turndown = await makeTurndown();
  return turndown.turndown(html);
}

async function markdownToHtml(markdown: string) {
  const { marked } = await import("marked");
  return marked(markdown);
}

function plainTextToHtml(source: string) {
  return escapeText(source).replace(/\n/g, "<br/>");
}

function htmlToPlaintext(html: string) {
  return htmlToMarkdown(html);
}

/**
 * end-consumer function to convert a note to HTML.
 */
export async function toHtml(format: NoteFormat, source: string) {
  let newHtml = "";
  if (format === NoteFormat.plain) {
    newHtml = plainTextToHtml(source);
  } else if (format === NoteFormat.markdown) {
    newHtml = await markdownToHtml(source);
  } else if (format === NoteFormat.richText) {
    newHtml = source;
  }
  const xss = await createXss();
  const xssed = xss.process(newHtml);
  const html = await enrichHtml(xssed);
  return html;
}

/**
 * convert a note from one format to another
 */
export async function convertNotes(
  oldFormat: NoteFormat,
  newFormat: NoteFormat,
  oldSource: string,
) {
  let newSource = "";
  let unsafeNewHtml = "";
  if (newFormat === NoteFormat.plain) {
    if (oldFormat === NoteFormat.markdown || oldFormat === NoteFormat.plain) {
      newSource = oldSource;
    } else if (oldFormat === NoteFormat.richText) {
      newSource = await htmlToPlaintext(oldSource);
    }
    unsafeNewHtml = plainTextToHtml(newSource);
  } else if (newFormat === NoteFormat.markdown) {
    if (oldFormat === NoteFormat.plain || oldFormat === NoteFormat.markdown) {
      newSource = oldSource;
    } else if (oldFormat === NoteFormat.richText) {
      newSource = await htmlToMarkdown(oldSource);
    }
    unsafeNewHtml = await markdownToHtml(newSource);
  } else if (newFormat === NoteFormat.richText) {
    if (oldFormat === NoteFormat.plain) {
      newSource = plainTextToHtml(oldSource);
    } else if (oldFormat === NoteFormat.markdown) {
      newSource = await markdownToHtml(oldSource);
    } else if (oldFormat === NoteFormat.richText) {
      newSource = oldSource;
    }
    unsafeNewHtml = newSource;
  }
  const xss = await createXss();
  const newHtml = await enrichHtml(xss.process(unsafeNewHtml));
  return { newSource, newHtml };
}
