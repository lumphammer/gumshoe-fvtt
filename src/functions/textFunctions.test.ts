import { describe, expect, test } from "vitest";

import { NoteFormat } from "../types";
import { convertNotes, toHtml } from "./textFunctions";

const markdownSample = `# This is a title

This is a paragraph. The next thing is a dangerous script tag:

## This is a subtitle

This is another paragraph. It has a [link](https://example.com) in it. It's quite long
and has a line break in it.

<script>alert("hello")</script>

This is a list:

- item 1
- item 2
- item 3
`;

describe("toHtml", async () => {
  const html = await toHtml("markdown", markdownSample);
  test("converts plain text to html", async () => {
    const result = await toHtml("plain", markdownSample);
    expect(result).toMatchSnapshot();
  });
  test("converts markdown to html", () => {
    expect(html).toMatchSnapshot();
  });
  test("converts rich text to html", async () => {
    const result = await toHtml("richText", html);
    expect(result).toMatchSnapshot();
  });
});

describe("convertNotes", async () => {
  const html = await toHtml("markdown", markdownSample);

  test.each<[NoteFormat, NoteFormat, string, string]>([
    ["plain", "plain", markdownSample, markdownSample],
    ["plain", "markdown", markdownSample, markdownSample],
    ["markdown", "plain", markdownSample, markdownSample],
    ["markdown", "markdown", markdownSample, markdownSample],
    ["richText", "richText", html, html],
  ])("converts %s to %s", async (oldFormat, newFormat, input, expected) => {
    const { newSource: result } = await convertNotes(
      oldFormat,
      newFormat,
      input,
    );
    expect(result).toBe(expected);
  });
  test.each<[NoteFormat, NoteFormat, string]>([
    ["plain", "richText", markdownSample],
    ["markdown", "richText", markdownSample],
    ["richText", "plain", html],
    ["richText", "markdown", html],
  ])("converts %s to %s", async (oldFormat, newFormat, input) => {
    const result = await convertNotes(oldFormat, newFormat, input);
    expect(result).toMatchSnapshot();
  });
});
