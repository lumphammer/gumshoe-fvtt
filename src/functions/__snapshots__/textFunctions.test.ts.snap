// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertNotes > converts markdown to richText 1`] = `
{
  "newHtml": "<h1>This is a title</h1>
<p>This is a paragraph. The next thing is a dangerous script tag:</p>
<h2>This is a subtitle</h2>
<p>This is another paragraph. It has a <a href="https://example.com">link</a> in it. It&#39;s quite long
and has a line break in it.</p>
&lt;script&gt;alert("hello")&lt;/script&gt;

<p>This is a list:</p>
<ul>
<li>item 1</li>
<li>item 2</li>
<li>item 3</li>
</ul>
",
  "newSource": "<h1>This is a title</h1>
<p>This is a paragraph. The next thing is a dangerous script tag:</p>
<h2>This is a subtitle</h2>
<p>This is another paragraph. It has a <a href="https://example.com">link</a> in it. It&#39;s quite long
and has a line break in it.</p>
<script>alert("hello")</script>

<p>This is a list:</p>
<ul>
<li>item 1</li>
<li>item 2</li>
<li>item 3</li>
</ul>
",
}
`;

exports[`convertNotes > converts plain to richText 1`] = `
{
  "newHtml": "# This is a title<br /><br />This is a paragraph. The next thing is a dangerous script tag:<br /><br />## This is a subtitle<br /><br />This is another paragraph. It has a [link](https://example.com) in it. It&#39;s quite long<br />and has a line break in it.<br /><br />&lt;script&gt;alert(&quot;hello&quot;)&lt;/script&gt;<br /><br />This is a list:<br /><br />- item 1<br />- item 2<br />- item 3<br />",
  "newSource": "# This is a title<br/><br/>This is a paragraph. The next thing is a dangerous script tag:<br/><br/>## This is a subtitle<br/><br/>This is another paragraph. It has a [link](https://example.com) in it. It&#39;s quite long<br/>and has a line break in it.<br/><br/>&lt;script&gt;alert(&quot;hello&quot;)&lt;/script&gt;<br/><br/>This is a list:<br/><br/>- item 1<br/>- item 2<br/>- item 3<br/>",
}
`;

exports[`convertNotes > converts richText to markdown 1`] = `
{
  "newHtml": "<h1>This is a title</h1>
<p>This is a paragraph. The next thing is a dangerous script tag:</p>
<h2>This is a subtitle</h2>
<p>This is another paragraph. It has a <a href="https://example.com">link</a> in it. It&#39;s quite long and has a line break in it.</p>
&lt;script&gt;alert("hello")&lt;/script&gt;

<p>This is a list:</p>
<ul>
<li>item 1</li>
<li>item 2</li>
<li>item 3</li>
</ul>
",
  "newSource": "This is a title
===============

This is a paragraph. The next thing is a dangerous script tag:

This is a subtitle
------------------

This is another paragraph. It has a [link](https://example.com) in it. It's quite long and has a line break in it.

<script>alert("hello")</script>

This is a list:

*   item 1
*   item 2
*   item 3",
}
`;

exports[`convertNotes > converts richText to plain 1`] = `
{
  "newHtml": "This is a title<br />===============<br /><br />This is a paragraph. The next thing is a dangerous script tag:<br /><br />This is a subtitle<br />------------------<br /><br />This is another paragraph. It has a [link](https://example.com) in it. It&#39;s quite long and has a line break in it.<br /><br />&lt;script&gt;alert(&quot;hello&quot;)&lt;/script&gt;<br /><br />This is a list:<br /><br />*   item 1<br />*   item 2<br />*   item 3",
  "newSource": "This is a title
===============

This is a paragraph. The next thing is a dangerous script tag:

This is a subtitle
------------------

This is another paragraph. It has a [link](https://example.com) in it. It's quite long and has a line break in it.

<script>alert("hello")</script>

This is a list:

*   item 1
*   item 2
*   item 3",
}
`;

exports[`toHtml > converts markdown to html 1`] = `
"<h1>This is a title</h1>
<p>This is a paragraph. The next thing is a dangerous script tag:</p>
<h2>This is a subtitle</h2>
<p>This is another paragraph. It has a <a href="https://example.com">link</a> in it. It&#39;s quite long
and has a line break in it.</p>
&lt;script&gt;alert("hello")&lt;/script&gt;

<p>This is a list:</p>
<ul>
<li>item 1</li>
<li>item 2</li>
<li>item 3</li>
</ul>
"
`;

exports[`toHtml > converts plain text to html 1`] = `"# This is a title<br /><br />This is a paragraph. The next thing is a dangerous script tag:<br /><br />## This is a subtitle<br /><br />This is another paragraph. It has a [link](https://example.com) in it. It&#39;s quite long<br />and has a line break in it.<br /><br />&lt;script&gt;alert(&quot;hello&quot;)&lt;/script&gt;<br /><br />This is a list:<br /><br />- item 1<br />- item 2<br />- item 3<br />"`;

exports[`toHtml > converts rich text to html 1`] = `
"<h1>This is a title</h1>
<p>This is a paragraph. The next thing is a dangerous script tag:</p>
<h2>This is a subtitle</h2>
<p>This is another paragraph. It has a <a href="https://example.com">link</a> in it. It&#39;s quite long
and has a line break in it.</p>
&lt;script&gt;alert("hello")&lt;/script&gt;

<p>This is a list:</p>
<ul>
<li>item 1</li>
<li>item 2</li>
<li>item 3</li>
</ul>
"
`;
